mm2src/coins/qrc20.rs

```diff
@@ -320,43 +340,40 @@ impl Qrc20Coin {
     /// Note: lock the UTXO_LOCK mutex before this function will be called.
     async fn generate_qrc20_transaction(
         &self,
-        outputs: Vec<ContractCallOutput>,
-    ) -> Result<GenerateQrc20TxResult, String> {
-        let unspents = try_s!(self
-            .ordered_mature_unspents(&self.utxo.my_address)
-            .compat()
-            .await
-            .map_err(|e| ERRL!("{}", e)));
+        contract_outputs: Vec<ContractCallOutput>,
+    ) -> Result<GenerateQrc20TxResult, GenerateTransactionError> {
+        let unspents = try_map!(
+            self.ordered_mature_unspents(&self.utxo.my_address).compat().await,
+            GenerateTransactionError::Other
+        );
 
-        // None seems that the generate_transaction() should request estimated fee for Kbyte
-        let actual_tx_fee = None;
-        let gas_fee = outputs
-            .iter()
-            .fold(0, |gas_fee, output| gas_fee + output.gas_limit * output.gas_price);
+        let mut gas_fee = 0;
+        let mut outputs = Vec::with_capacity(contract_outputs.len());
+        for output in contract_outputs {
+            gas_fee += output.gas_limit * output.gas_price;
+            outputs.push(TransactionOutput::from(output));
+        }
         let fee_policy = FeePolicy::SendExact;
+        let tx_fee = None;
 
-        let outputs = outputs.into_iter().map(|output| output.into()).collect();
         let (unsigned, data) = self
-            .generate_transaction(unspents, outputs, fee_policy, actual_tx_fee, Some(gas_fee))
-            .await
-            .map_err(|e| match &e {
-                GenerateTransactionError::EmptyUtxoSet => ERRL!("Not enough {} to Pay Fee: {}", self.platform, e),
-                GenerateTransactionError::NotSufficientBalance { description } => {
-                    ERRL!("Not enough {} to Pay Fee: {}", self.platform, description)
-                },
-                e => ERRL!("{}", e),
-            })?;
+            .generate_transaction(unspents, outputs, fee_policy, tx_fee, Some(gas_fee))
+            .await?;
         let prev_script = ScriptBuilder::build_p2pkh(&self.utxo.my_address.hash);
-        let signed = try_s!(sign_tx(
-            unsigned,
-            &self.utxo.key_pair,
-            prev_script,
-            self.utxo.signature_version,
-            self.utxo.fork_id
-        ));
+        let signed = try_map!(
+            sign_tx(
+                unsigned,
+                &self.utxo.key_pair,
+                prev_script,
+                self.utxo.signature_version,
+                self.utxo.fork_id
+            ),
+            GenerateTransactionError::Other
+        );
+        let miner_fee = data.fee_amount + data.unused_change.unwrap_or_default();
```

Comment:
It is now clear after discussion in person. The unused_change is not added during trade preimage because it can be included in the available trade volume: https://github.com/KomodoPlatform/atomicDEX-API/blob/mm2.1-refactor-get-trade-fee/mm2src/coins/utxo/utxo_common.rs#L1774